### DISCUSSION TITLE
Using quarkus with older version of keycloak admin client

### DISCUSSION THREAD

**Discussion by sbergandy** [quarkusio/quarkus#36766] [Community]
Hi

how can I set up pom.xml in order to use quarkus 3.5.0 with keycload-admin-client 20.0.3? Many thanks in advance.

Cheers
Sebastian

**Comment by maxandersen** [comment]
You override the dependencies. But not sure if that old adm client is possible. Question for keycloak what versions they cover.

**Reply by sbergandy** [reply]
I'm trying to override the dependencies. The main problem I have currently is that Quarkus 3 is based on Jakarta EE 10 (which is good) and Keycloak 20.0.3 on Java EE. When communicating with Keycloak I'm getting exception like:
```java
Caused by: java.lang.ClassNotFoundException: javax.ws.rs.client.ClientBuilder
```
I've added the dependency:
```xml
		<dependency>
			<groupId>javax.ws.rs</groupId>
			<artifactId>javax.ws.rs-api</artifactId>
			<version>${javax.ws.rs-api}</version>
		</dependency>
```
in order to make the missing class available, but still without success.
The main question is: is it possible to use Quarkus 3 Jakarta EE 10 based together with older Java EE based keycloak admin client?
I think this is a general case when someone write a Quarkus application and would like to communicate with older Keycloak server.

**Reply by maxandersen** [reply]
You can't use libraries that use the old javax or rather it's going to be tricky. 

Keycloack clients should be fairly backwards compatible. But not sure about the admin client. 

Question is best answered by keycloak team.

**Reply by sberyozkin** [reply]
@sbergandy Does quarkus-keycloak-admin-client shipped with Quarkus 3.5.0 not work against KC 20.0.3 ? Can you clarify please why you are trying to manage keycloak-admin-client dependency,  thanks

**Reply by sbergandy** [reply]
I found my problem right now. The `org.keycloak.representations.idm.UserRepresentation` has new field `userProfileMetadata` which is sent with `PUT` request as within JSON when updating user with `org.keycloak.admin.client.resource.UserResource#update`. The Keycloak server 20.0.3 anserws with 400. Keycloak server logs `[org.keycloak.services.resources.admin.UserResource] (executor-thread-105553) Could not update user!: java.lang.NullPointerException` . I tried to add an empty `UserProfileMetada` object:
```json
  "userProfileMetadata": {
    "attributes": null
  }
```
but then I got in response body messasge: `"error": "Unrecognized field \"userProfileMetadata\" (class org.keycloak.representations.idm.UserRepresentation), not marked as ignorable (26 known properties:...`
Right know I'm trying to exclude this field from being serialized into JSON. I don't know if that possible/configurable?

**Reply by sberyozkin** [reply]
Thanks for tracing the problem. I'm not sure we can control it at the level of keycloak admin client calls as opposed to the regular rest client calls, as far as I recall keyclok-admin-client does not use default Jackson provider. I can have a better look but only next week when I can get to my laptop,  CC-ing to @geoand and @michalvavrik in case they might recall something

**Reply by sbergandy** [reply]
I managed to make it work by extending the `org.keycloak.representations.idm.UserRepresentation` and ignoring unwanted fields like this:
```java
class OldKeycloakUser extends UserRepresentation {

	@JsonIgnore
	private UserProfileMetadata userProfileMetadata;
```
Once we will start to use the newer Keycloak server version I can remove it. I think this should be fine IMO.

**Reply by sberyozkin** [reply]
Cool,  great stuff 👍

**Reply by michalvavrik** [reply]
It's fine @sbergandy as long as you know what you are doing.

I think we tried to prevent users from ignoring unknown fields as we adapt Keycloak Admin Client to specific Keycloak version and if you are using it with different Keycloak version, we can't guarantee it will work (native mode for one). Here @geoand tried to make sure that you can't change `quarkus.jackson.fail-on-unknown-properties=false`, however unless I'm very wrong, he didn't do it right - it checks for `propertyName.startsWith("io.quarkus.jackson")` while I think it should be `propertyName.startsWith("quarkus.jackson")`. I'll create PR that corrects it, maybe I'm simply missing something.

https://github.com/quarkusio/quarkus/blob/main/extensions/keycloak-admin-client-reactive/runtime/src/main/java/io/quarkus/keycloak/admin/client/reactive/runtime/ResteasyReactiveClientProvider.java#L102

**Reply by michalvavrik** [reply]
Hmm, now that I'm thinking about it, you are using Keycloak Admin Client (not Reactive version), that is why you are experiencing it as it has different default client setup.

**Reply by sbergandy** [reply]
Yep, I'm not using reactive version. What can go wrong when sending JSON without one field according to expected format?

**Reply by michalvavrik** [reply]
> Yep, I'm not using reactive version. What can go wrong when sending JSON without one field according to expected format?

That's the point, I don't know. Can you really tell if some field is vital without going through Keycloak code first? Just to be clear, I don't suggest your changes are wrong as long as you know what you do :-) No objections to your workaround for your case!

**Reply by sbergandy** [reply]
Thanks for your time and anserws.

**Comment by sberyozkin** [comment]
Thanks @michalvavrik for the PR fixing a typo in the reactive version. 
IMHO the workaround in this case is ok, it simply fixes a backward compatibility issue, blocking an unrecognised field

**Reply by sbergandy** [reply]
Thanks. This was exactly my intention i.e. simply fix for backward compatibility. If Keycloak 20.0.3 doesn't expect a field because it will be first added in a newer version then I'm not sending it. Once we upgrade our server the newer version I can just remove the class. I also tested Keycloak via curl by putting the user representation without and with `userProfileMetada` field.

**Comment by sbergandy** [comment]
Thank you Guys for your time.

**Comment by vmuzikar** [comment]
@sbergandy Could you please create an issue in the [Keycloak repo](https://github.com/keycloak/keycloak) around a potential backwards compatibility issue?

**Reply by sbergandy** [reply]
I think this isn't an issue because they don't guarantee that in case of version mismatch everything will work. See https://www.keycloak.org/docs/latest/upgrading/index.html

![image](https://github.com/quarkusio/quarkus/assets/1845106/fab5693b-e82e-44e0-b297-1fdc4664cdeb)

What do you think?

**Reply by vmuzikar** [reply]
We'd still like to investigate and see if there's anything we can do about it. :) This is connected to the cyclic dependency issue we have with Quarkus.

**Reply by sbergandy** [reply]
Ok I will create an issue, but I think we can't do anything because this is by design so. Keycloak answers with 400 if unknown field was sent in UserRepresentation JSON.

**Reply by sbergandy** [reply]
done https://github.com/keycloak/keycloak/issues/24393
